### ISSUE TITLE
input csv file format; must use salae software, or are foss alternatives usable?

### ISSUE THREAD

**Issue by hanetzer** [PSPReverse/PSPTrace#2]
As the topic. I'm not too familiar with the salae ecosystem, but I have had a passing acquaintance with
sigrok/pulseview. Could an export of the trace from said software be used, or would this require some
alterations of the psptrace tool as it currently exists?

**Comment by hanetzer** [comment]
If my reading is correct, you only can support salae produced data. If so, could someone produce
said data, and provide the mobo name/bios version so I could see about badgering sigrok support
into the project?

**Comment by cwerling** [comment]
Correct, PSPTrace only works with Saleae Logic's CSV output format. However, it shouldn't be too hard to either adapt sigrok's output format or build some additional parsing logic into PSPTrace. In fact, that would probably be helpful for other people who('s organisation) cannot afford the Saleae Logic Analyzers or prefer an open source tool anyhow.

The CSV output format of Saleae looks like this:

```csv
Time [s],Packet ID,MOSI,MISO
-9.761530688000001,8,0x03,0xFF
-9.761530208000000,8,0x02,0xFF
-9.761529727999999,8,0x00,0xFF
-9.761529248000000,8,0x00,0xFF
-9.761528768000000,8,0x00,0xAA
-9.761528287999999,8,0x00,0x55
-9.761527808000000,8,0x00,0xAA
-9.761527328000000,8,0x00,0x55
```

This is a redacted example, as these files usually get very big. You can see here, e.g., how the `0x03` mainboard master queries a read to `0x200000` (big endian) and then receives the Firmware Entry Table magic `0x55AA55AA`.

Hope this helps!

**Comment by hanetzer** [comment]
Ah, this seems fairly reasonable. However, I'm not sure what to make of the 'packet id' field.

**Comment by hanetzer** [comment]
seems current saleae software has changed the format, or perhaps I'm just doing it wrong:
```
name,type,start_time,duration,"mosi","miso"
"SPI","enable",2.75874658,4.16666671e-08,,
"SPI","result",2.75874667,1.08333333e-06,0x1E,0xFF
"SPI","result",2.75874775,1.08333333e-06,0x00,0xFF
"SPI","result",2.75874883,1.08333333e-06,0x00,0xFF
"SPI","result",2.75874996,1.04166667e-06,0x00,0xFF
"SPI","result",2.75875104,1.04166667e-06,0x00,0xFF
"SPI","result",2.75875212,1.04166667e-06,0x00,0xFF
"SPI","result",2.75875321,1.08333333e-06,0x00,0xFF
"SPI","disable",2.75875488,4.16666667e-08,,
"SPI","enable",2.75875504,4.16666667e-08,,
"SPI","result",2.75875512,1.04166667e-06,0x1E,0xFF
"SPI","result",2.75875621,1.08333333e-06,0x00,0xFF
"SPI","result",2.75875729,1.08333333e-06,0x00,0xFF
"SPI","result",2.75875842,1.04166667e-06,0x00,0xFF
"SPI","result",2.7587595,1.04166667e-06,0x00,0xFF
```

**Comment by hanetzer** [comment]
ah nvm, I was exporting it incorrectly.

**Comment by hanetzer** [comment]
I will say, however, I appear to be missing data in some form or another. I've attempted to use
coreboot's `util/amdtools/update_efs_spi_speed` to 'slow down' the whole process so my saleae
knockoff I forgot I had, which can only do 24mhz sampling at fastest, would be able to capture
a complete/accurate dump. The machine boots, but I never see it grab `0x55AA55AA` off the chip.
it must be doing it, otherwise it wouldn't boot, no?

**Comment by cwerling** [comment]
Can you provide a screenshot of the logic trace so we can reason about its correctness?

**Comment by hanetzer** [comment]
exactly what do you mean, the saleae software, the exported csv, or the results from psptrace?

**Comment by cwerling** [comment]
I mean from the (Saelae) Logic software. How do the very first recorded logic signals look like?

**Comment by hanetzer** [comment]
sec, I'll have to redo the cap; I didn't save the raw data.

**Comment by hanetzer** [comment]
![logic](https://github.com/PSPReverse/PSPTrace/assets/2760645/7819dfe1-4a98-46ab-b526-5b83a0c141f5)
here's the very start of it.

**Comment by cwerling** [comment]
If 0x1E is really supposed to be the first byte (i.e. instruction) from the master, this doesn't look correct to me, as I couldn't find 0x1E to be a valid instruction in a [Winbond manual](https://www.mouser.com/datasheet/2/949/W25R256JV_RevD_06162020-1879221.pdf). I would suspect either wrong SPI decode settings in Logic or a bad sample (too low sample rate, wrong voltage threshold, wonky connections).

Try zooming in on the very first MOSI part. It should be some kind of read, going through the [known FET addresses](https://github.com/PSPReverse/PSPTool/blob/master/psptool/blob.py#L41) and looking for the [FET signature](https://github.com/PSPReverse/PSPTool/blob/master/psptool/blob.py#L29). At some point MISO should reply with the signature.

**Comment by hanetzer** [comment]
well, my knockoff saleae can only manage 24Mhz, so maybe its just not fast enough for nyquist to be happy.
However, now that I know the format expected by this software, I can probably work something out with
sigrok and the kingst logic analyzer i intend on getting in the hopefully near future.

**Comment by hanetzer** [comment]
Got a kingst la2016 from a kind soul (their words: sitting on a shelf collecting dust).
Managed to get a good bootup trace. 
also, a user by the name of gsi on the sigrok irc channel put together a nice awk script
to turn a serialice based dump into one compatible with your software:

```awk
#!/usr/bin/awk -f
# Feed psptrace from sigrok-cli SPI decoder output.
# Desired output text format:
#
#   Tims [s],Packet ID,MOSI,MISO
#   8.5,8,0x03,0xff
#
# Example invocation:
#
#   (get the samplerate)
#   $ sigrok-cli -i $F --show
#
#   (decode SPI,and feed the converter)
#   $ sigrok-cli -i $F \
#       -P spi:cs=CS:clk=CLK:mosi=MOSI:miso=MISO \
#       -A spi=mosi-data:miso-data \
#       --protocol-decoder-ann-class --protocol-decoder-samplenum \
#     awk -v RATE=1000 -f spi-dump-conv.awk > spi-dump.csv

BEGIN {
	print("Tims [s],Packet ID,MOSI,MISO")
}

$3 ~ /miso-data/ {
	MISO = $4
	next
}

$3 ~ /mosi-data/ {
	MOSI = $4
	TS = $1
	sub("^.*-","",TS)
	TS /= RATE
	printf("%f,%d,0x%s,0x%s\n",TS,8,MOSI,MISO)
}
```
